File: .claude/skills/scenario-simulation/templates/scenario_simulator.ts
/**
 * Scenario Simulation Template (TypeScript)
 * A flexible framework for running scenario-based simulations in TypeScript/JavaScript
 */

interface VariableConfig {
  name: string;
  bestCase: number | string;
  expectedCase: number | string;
  worstCase: number | string;
  description?: string;
  unit?: string;
}

interface ScenarioInputs {
  [key: string]: number | string;
}

interface ScenarioResults {
  [key: string]: number | string;
}

type ModelFunction = (inputs: ScenarioInputs) => ScenarioResults;
type ScenarioType = 'best_case' | 'expected_case' | 'worst_case';

class Variable {
  name: string;
  bestCase: number | string;
  expectedCase: number | string;
  worstCase: number | string;
  description: string;
  unit: string;

  constructor(config: VariableConfig) {
    this.name = config.name;
    this.bestCase = config.bestCase;
    this.expectedCase = config.expectedCase;
    this.worstCase = config.worstCase;
    this.description = config.description || '';
    this.unit = config.unit || '';
  }

  getValue(scenario: ScenarioType): number | string {
    const scenarios = {
      best_case: this.bestCase,
      expected_case: this.expectedCase,
      worst_case: this.worstCase,
    };
    return scenarios[scenario];
  }
}

class ScenarioSimulator {
  private variables: Map<string, Variable>;
  private results: Map<string, ScenarioResults>;

  constructor(variables: Variable[]) {
    this.variables = new Map(variables.map(v => [v.name, v]));
    this.results = new Map();
  }

  /**
   * Get all input values for a specific scenario
   */
  getInputs(scenario: ScenarioType): ScenarioInputs {
    const inputs: ScenarioInputs = {};
    this.variables.forEach((variable, name) => {
      inputs[name] = variable.getValue(scenario);
    });
    return inputs;
  }

  /**
   * Run a single scenario
   */
  runScenario(scenario: ScenarioType, modelFunc: ModelFunction): ScenarioResults {
    const inputs = this.getInputs(scenario);
    const result = modelFunc(inputs);
    this.results.set(scenario, result);
    return result;
  }

  /**
   * Run all standard scenarios
   */
  runAllScenarios(modelFunc: ModelFunction): Map<string, ScenarioResults> {
    const scenarios: ScenarioType[] = ['best_case', 'expected_case', 'worst_case'];
    scenarios.forEach(scenario => {
      this.runScenario(scenario, modelFunc);
    });
    return this.results;
  }

  /**
   * Run a custom scenario with specific input values
   */
  runCustomScenario(
    scenarioName: string,
    customInputs: ScenarioInputs,
    modelFunc: ModelFunction
  ): ScenarioResults {
    const result = modelFunc(customInputs);
    this.results.set(scenarioName, result);
    return result;
  }

  /**
   * Perform sensitivity analysis on a single variable
   */
  sensitivityAnalysis(
    variableName: string,
    modelFunc: ModelFunction,
    numSteps: number = 10
  ): Array<ScenarioResults & { [key: string]: number }> {
    const variable = this.variables.get(variableName);
    if (!variable) {
      throw new Error(`Variable ${variableName} not found`);
    }

    const baseInputs = this.getInputs('expected_case');
    const results: Array<ScenarioResults & { [key: string]: number }> = [];

    // Check if variable is numeric
    if (typeof variable.expectedCase === 'number') {
      const minVal = Math.min(
        variable.worstCase as number,
        variable.bestCase as number
      );
      const maxVal = Math.max(
        variable.worstCase as number,
        variable.bestCase as number
      );
      const step = (maxVal - minVal) / (numSteps - 1);

      for (let i = 0; i < numSteps; i++) {
        const value = minVal + step * i;
        const inputs = { ...baseInputs, [variableName]: value };
        const result = modelFunc(inputs);
        results.push({ [variableName]: value, ...result });
      }
    } else {
      // For non-numeric variables, test the three defined values
      [variable.worstCase, variable.expectedCase, variable.bestCase].forEach(value => {
        const inputs = { ...baseInputs, [variableName]: value };
        const result = modelFunc(inputs);
        results.push({ [variableName]: value as number, ...result });
      });
    }

    return results;
  }

  /**
   * Run Monte Carlo simulation with random sampling
   */
  monteCarloSimulation(
    modelFunc: ModelFunction,
    nIterations: number = 1000,
    randomSeed?: number
  ): ScenarioResults[] {
    // Simple seeded random number generator
    let seed = randomSeed || Date.now();
    const random = () => {
      seed = (seed * 9301 + 49297) % 233280;
      return seed / 233280;
    };

    const results: ScenarioResults[] = [];

    for (let i = 0; i < nIterations; i++) {
      const inputs: ScenarioInputs = {};

      this.variables.forEach((variable, name) => {
        if (typeof variable.expectedCase === 'number') {
          // Triangular distribution
          const worst = variable.worstCase as number;
          const expected = variable.expectedCase as number;
          const best = variable.bestCase as number;

          const u = random();
          const fc = (expected - worst) / (best - worst);

          let value: number;
          if (u < fc) {
            value = worst + Math.sqrt(u * (best - worst) * (expected - worst));
          } else {
            value = best - Math.sqrt((1 - u) * (best - worst) * (best - expected));
          }
          inputs[name] = value;
        } else {
          inputs[name] = variable.expectedCase;
        }
      });

      results.push(modelFunc(inputs));
    }

    return results;
  }

  /**
   * Create comparison table of all scenarios
   */
  compareScenarios(): { metric: string; [scenario: string]: number | string }[] {
    if (this.results.size === 0) {
      throw new Error('No results to compare. Run scenarios first.');
    }

    // Get all metrics from first result
    const firstResult = Array.from(this.results.values())[0];
    const metrics = Object.keys(firstResult);

    return metrics.map(metric => {
      const row: { metric: string; [scenario: string]: number | string } = { metric };
      this.results.forEach((results, scenario) => {
        row[scenario] = results[metric] ?? 'N/A';
      });
      return row;
    });
  }

  /**
   * Export results to JSON
   */
  exportResults(): string {
    const resultsObj: { [key: string]: ScenarioResults } = {};
    this.results.forEach((value, key) => {
      resultsObj[key] = value;
    });
    return JSON.stringify(resultsObj, null, 2);
  }

  /**
   * Print formatted summary
   */
  printSummary(): void {
    console.log('\n' + '='.repeat(60));
    console.log('SCENARIO SIMULATION SUMMARY');
    console.log('='.repeat(60));

    console.log('\nInput Variables:');
    this.variables.forEach((variable, name) => {
      console.log(`\n  ${name}:`);
      console.log(`    Best Case: ${variable.bestCase} ${variable.unit}`);
      console.log(`    Expected:  ${variable.expectedCase} ${variable.unit}`);
      console.log(`    Worst Case: ${variable.worstCase} ${variable.unit}`);
    });

    if (this.results.size > 0) {
      console.log('\n' + '-'.repeat(60));
      console.log('Results Comparison:');
      console.log('-'.repeat(60));
      console.table(this.compareScenarios());
    }

    console.log('\n' + '='.repeat(60) + '\n');
  }

  /**
   * Calculate statistics from Monte Carlo results
   */
  static calculateStatistics(results: ScenarioResults[]): {
    [metric: string]: {
      mean: number;
      median: number;
      std: number;
      min: number;
      max: number;
      p5: number;
      p95: number;
    };
  } {
    if (results.length === 0) return {};

    const metrics = Object.keys(results[0]);
    const stats: any = {};

    metrics.forEach(metric => {
      const values = results
        .map(r => r[metric])
        .filter(v => typeof v === 'number')
        .sort((a, b) => (a as number) - (b as number)) as number[];

      if (values.length === 0) return;

      const sum = values.reduce((a, b) => a + b, 0);
      const mean = sum / values.length;
      const variance = values.reduce((a, b) => a + Math.pow(b - mean, 2), 0) / values.length;
      const std = Math.sqrt(variance);

      stats[metric] = {
        mean: Math.round(mean * 100) / 100,
        median: values[Math.floor(values.length / 2)],
        std: Math.round(std * 100) / 100,
        min: values[0],
        max: values[values.length - 1],
        p5: values[Math.floor(values.length * 0.05)],
        p95: values[Math.floor(values.length * 0.95)],
      };
    });

    return stats;
  }
}

// Example usage
if (require.main === module) {
  // Define variables
  const variables = [
    new Variable({
      name: 'monthly_users',
      bestCase: 10000,
      expectedCase: 5000,
      worstCase: 1000,
      description: 'Number of monthly active users',
      unit: 'users',
    }),
    new Variable({
      name: 'conversion_rate',
      bestCase: 0.1,
      expectedCase: 0.05,
      worstCase: 0.02,
      description: 'Percentage of users who convert',
      unit: '%',
    }),
    new Variable({
      name: 'avg_revenue_per_user',
      bestCase: 100,
      expectedCase: 50,
      worstCase: 20,
      description: 'Average revenue per paying user',
      unit: '$',
    }),
    new Variable({
      name: 'monthly_costs',
      bestCase: 50000,
      expectedCase: 75000,
      worstCase: 100000,
      description: 'Monthly operating costs',
      unit: '$',
    }),
  ];

  // Define model function
  const revenueModel: ModelFunction = (inputs) => {
    const payingUsers = (inputs.monthly_users as number) * (inputs.conversion_rate as number);
    const revenue = payingUsers * (inputs.avg_revenue_per_user as number);
    const costs = inputs.monthly_costs as number;
    const profit = revenue - costs;
    const roi = costs > 0 ? (profit / costs) * 100 : 0;

    return {
      paying_users: Math.round(payingUsers),
      monthly_revenue: Math.round(revenue * 100) / 100,
      monthly_costs: Math.round(costs * 100) / 100,
      monthly_profit: Math.round(profit * 100) / 100,
      roi_percent: Math.round(roi * 100) / 100,
    };
  };

  // Run simulation
  const simulator = new ScenarioSimulator(variables);

  console.log('Running scenario analysis...');
  simulator.runAllScenarios(revenueModel);
  simulator.printSummary();

  // Sensitivity analysis
  console.log('\nRunning sensitivity analysis on conversion_rate...');
  const sensitivityResults = simulator.sensitivityAnalysis('conversion_rate', revenueModel);
  console.table(sensitivityResults);

  // Monte Carlo simulation
  console.log('\nRunning Monte Carlo simulation (1000 iterations)...');
  const mcResults = simulator.monteCarloSimulation(revenueModel, 1000, 42);

  console.log('\nMonte Carlo Results Summary:');
  const stats = ScenarioSimulator.calculateStatistics(mcResults);
  console.table(stats);

  // Export results
  console.log('\nResults JSON:');
  console.log(simulator.exportResults());
}

// Export for use as a module
export { ScenarioSimulator, Variable, ScenarioInputs, ScenarioResults, ModelFunction };

File: .claude/skills/scenario-simulation/templates/scenario_simulator.py
#!/usr/bin/env python3
"""
Scenario Simulation Template
A flexible framework for running scenario-based simulations
"""

import pandas as pd
import numpy as np
from typing import Dict, List, Any, Callable
from dataclasses import dataclass
import json


@dataclass
class Variable:
    """Represents a simulation variable with different scenario values"""
    name: str
    best_case: Any
    expected_case: Any
    worst_case: Any
    description: str = ""
    unit: str = ""

    def get_value(self, scenario: str) -> Any:
        """Get the value for a specific scenario"""
        scenarios = {
            'best_case': self.best_case,
            'expected_case': self.expected_case,
            'worst_case': self.worst_case
        }
        return scenarios.get(scenario, self.expected_case)


class ScenarioSimulator:
    """Base class for scenario simulations"""

    def __init__(self, variables: List[Variable]):
        self.variables = {var.name: var for var in variables}
        self.results = {}
        self.metrics = {}

    def get_inputs(self, scenario: str) -> Dict[str, Any]:
        """Get all input values for a specific scenario"""
        return {
            name: var.get_value(scenario)
            for name, var in self.variables.items()
        }

    def run_scenario(self, scenario: str, model_func: Callable) -> Dict[str, Any]:
        """
        Run a single scenario

        Args:
            scenario: 'best_case', 'expected_case', or 'worst_case'
            model_func: Function that takes inputs dict and returns results dict
        """
        inputs = self.get_inputs(scenario)
        result = model_func(inputs)
        self.results[scenario] = result
        return result

    def run_all_scenarios(self, model_func: Callable) -> Dict[str, Dict[str, Any]]:
        """Run all standard scenarios"""
        scenarios = ['best_case', 'expected_case', 'worst_case']
        for scenario in scenarios:
            self.run_scenario(scenario, model_func)
        return self.results

    def run_custom_scenario(self, scenario_name: str, custom_inputs: Dict[str, Any],
                           model_func: Callable) -> Dict[str, Any]:
        """Run a custom scenario with specific input values"""
        result = model_func(custom_inputs)
        self.results[scenario_name] = result
        return result

    def sensitivity_analysis(self, variable_name: str, model_func: Callable,
                           num_steps: int = 10) -> pd.DataFrame:
        """
        Perform sensitivity analysis on a single variable

        Args:
            variable_name: Name of variable to analyze
            model_func: Model function to run
            num_steps: Number of values to test
        """
        if variable_name not in self.variables:
            raise ValueError(f"Variable {variable_name} not found")

        var = self.variables[variable_name]
        base_inputs = self.get_inputs('expected_case')

        # Create range of values
        if isinstance(var.expected_case, (int, float)):
            min_val = min(var.worst_case, var.best_case)
            max_val = max(var.worst_case, var.best_case)
            values = np.linspace(min_val, max_val, num_steps)
        else:
            # For non-numeric variables, just test the three defined values
            values = [var.worst_case, var.expected_case, var.best_case]

        results = []
        for value in values:
            inputs = base_inputs.copy()
            inputs[variable_name] = value
            result = model_func(inputs)
            results.append({
                variable_name: value,
                **result
            })

        return pd.DataFrame(results)

    def monte_carlo_simulation(self, model_func: Callable,
                              n_iterations: int = 1000,
                              random_seed: int = 42) -> pd.DataFrame:
        """
        Run Monte Carlo simulation with random sampling

        Assumes triangular distribution between worst, expected, and best cases
        """
        np.random.seed(random_seed)
        results = []

        for _ in range(n_iterations):
            inputs = {}
            for name, var in self.variables.items():
                if isinstance(var.expected_case, (int, float)):
                    # Triangular distribution
                    value = np.random.triangular(
                        var.worst_case,
                        var.expected_case,
                        var.best_case
                    )
                    inputs[name] = value
                else:
                    # For non-numeric, use expected case
                    inputs[name] = var.expected_case

            result = model_func(inputs)
            results.append(result)

        return pd.DataFrame(results)

    def compare_scenarios(self, metrics: List[str] = None) -> pd.DataFrame:
        """Create comparison table of all scenarios"""
        if not self.results:
            raise ValueError("No results to compare. Run scenarios first.")

        if metrics is None:
            # Use all keys from first result
            metrics = list(next(iter(self.results.values())).keys())

        comparison = {}
        for metric in metrics:
            comparison[metric] = {
                scenario: results.get(metric, 'N/A')
                for scenario, results in self.results.items()
            }

        return pd.DataFrame(comparison).T

    def export_results(self, filepath: str, format: str = 'json'):
        """Export results to file"""
        if format == 'json':
            with open(filepath, 'w') as f:
                json.dump(self.results, f, indent=2, default=str)
        elif format == 'csv':
            df = self.compare_scenarios()
            df.to_csv(filepath)
        else:
            raise ValueError(f"Unsupported format: {format}")

    def print_summary(self):
        """Print a formatted summary of results"""
        print("\n" + "="*60)
        print("SCENARIO SIMULATION SUMMARY")
        print("="*60)

        print("\nInput Variables:")
        for name, var in self.variables.items():
            print(f"\n  {name}:")
            print(f"    Best Case: {var.best_case} {var.unit}")
            print(f"    Expected:  {var.expected_case} {var.unit}")
            print(f"    Worst Case: {var.worst_case} {var.unit}")

        if self.results:
            print("\n" + "-"*60)
            print("Results Comparison:")
            print("-"*60)
            df = self.compare_scenarios()
            print(f"\n{df.to_string()}")

        print("\n" + "="*60 + "\n")


# Example usage
if __name__ == "__main__":
    # Define variables
    variables = [
        Variable(
            name="monthly_users",
            best_case=10000,
            expected_case=5000,
            worst_case=1000,
            description="Number of monthly active users",
            unit="users"
        ),
        Variable(
            name="conversion_rate",
            best_case=0.10,
            expected_case=0.05,
            worst_case=0.02,
            description="Percentage of users who convert",
            unit="%"
        ),
        Variable(
            name="avg_revenue_per_user",
            best_case=100,
            expected_case=50,
            worst_case=20,
            description="Average revenue per paying user",
            unit="$"
        ),
        Variable(
            name="monthly_costs",
            best_case=50000,
            expected_case=75000,
            worst_case=100000,
            description="Monthly operating costs",
            unit="$"
        )
    ]

    # Define model function
    def revenue_model(inputs: Dict[str, Any]) -> Dict[str, Any]:
        """Simple revenue model"""
        paying_users = inputs['monthly_users'] * inputs['conversion_rate']
        revenue = paying_users * inputs['avg_revenue_per_user']
        costs = inputs['monthly_costs']
        profit = revenue - costs
        roi = (profit / costs * 100) if costs > 0 else 0

        return {
            'paying_users': int(paying_users),
            'monthly_revenue': round(revenue, 2),
            'monthly_costs': round(costs, 2),
            'monthly_profit': round(profit, 2),
            'roi_percent': round(roi, 2)
        }

    # Run simulation
    simulator = ScenarioSimulator(variables)

    print("Running scenario analysis...")
    simulator.run_all_scenarios(revenue_model)
    simulator.print_summary()

    # Sensitivity analysis
    print("\nRunning sensitivity analysis on conversion_rate...")
    sensitivity_df = simulator.sensitivity_analysis('conversion_rate', revenue_model)
    print(sensitivity_df.to_string(index=False))

    # Monte Carlo simulation
    print("\nRunning Monte Carlo simulation (1000 iterations)...")
    mc_results = simulator.monte_carlo_simulation(revenue_model, n_iterations=1000)

    print("\nMonte Carlo Results Summary:")
    print(mc_results.describe())

    # Export results
    simulator.export_results('scenario_results.json', format='json')
    print("\nResults exported to scenario_results.json")
